### PR TITLE
Introducing security best practices doc

### PR DESCRIPTION
**Pull request by douglasmaciver**
A small, one-file commit that introduces best practices for keeping
enclaves secure. Use this guide to apply secure patterns and
avoid common mistakes that put the security of enclave
applications at risk.

Credit for early draft work to mingweishih.

Fixes openenclave#4031

Signed-off-by: Douglas MacIver <domac@microsoft.com>

### CODE REVIEW

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:106]:
```suggestion
2)	The bounding arithmetic operations do not wrap (i.e., numerical overflow).
```
nit: always put comma after "i.e." per Chicago Manual of Style

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:77]: This section is not labeled as SGX specific, but the discussion appears to make an SGX-specific assumption that host memory can be directly read by the enclave.  This is not true in general, where line 94 below is unsafe because the memory pointed to by blob might not be readable from the encave.   If this intends to talk specifically about SGX again, then it should say so, like section 1.2 does.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:176]:
```suggestion
> A warning about legacy data-parsers: Take care when passing "hot" data to code that may not have been written to parse maliciously crafted input. We have seen cases where legacy code that parses self-formatting or self-referencing data was used in new enclave applications. This can lead to significant vulnerabilities if the data being parsed is still controlled by the host, as may be the case with custom marshaling. In the case of this sample code, as mentioned earlier, `render_data()` parses "flat" data that is not nested nor self-referencing, so it's safe (and performant) to pass it "hot" data that is controlled by the host. Care must be taken though, to avoid data-consistency and other app-level integrity problems.
```

also, split source lines into shorter lines.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:209]: It's only corrected for SGX, not for other TEEs because line 224 etc still assumes it can read host memory.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]:
```suggestion
Open Enclave is an SDK that helps developers build apps that will run inside a hardware-based Trusted Execution Environment (TEE). At their core, TEEs protect application code and data at runtime from the host environment. Without new, hardware-implemented protections, a malicious or compromised host operating system would be able to modify code or read data. Any secrets managed by an application during runtime would be at risk of exposure. The Open Enclave SDK helps developers build applications that are protected from these threats.
```

Grammar nit.  Also again, please split long lines into multiple lines in the source file.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:248]:
```suggestion
> It's important to note that enclaves do not generally provide _code confidentiality_. This is noteworthy in the context of this discussion because, as will be discussed, hard-coded credentials in application code are to be avoided -- with or without enclave protections.
```

Add "generally" since there are ways to do it by loading an encrypted code module into an enclave.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:271]:
```suggestion
The enclave code then calculates a hash (e.g., SHA256) value of the public key and passes it as an input to code which generates attestation evidence. For Open Enclave, this can be done with `oe_get_evidence`.
```
Avoid SGX-specific terminology and references to deprecated/removed APIs.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:273]:
```suggestion
The client then sends the evidence and the serialized key to an attestation service. The attestation service will appraise the evidence and ensure that the hash of the key is present in the evidence and will issue an Attestation Result (sometimes called an "Attestation Token").
```

I'd prefer to delete "Attestation Token" as a capitalized term as Attestation Result is the standard term, and "Attestation Token" does not appear in any OE header or other document.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:275]: This statement is confusing.  Why would the Attestation Result need to contain the serialized key?   The serialized key can be obtained or negotiated via the protocol between the attester and the relying party once the RP accepts the attestation token.  Why would you want to reveal the serialized key to the attestation service?   This seems less secure (doesn't follow the need to know principle.)   Given this is a security best practices document, I think any recommendation to reveal the key to the attestation service should be removed.

**Review by dthaler — DISMISSED** [review]

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:41]: Good catch on boilerplate. Updated.
Limited Markdown text lines to 80, with the exception of links.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:106]: Yes! The Chicago Manual of Style.
The Microsoft Style Guide also states not to use "i.e.": https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/i/ie

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:77]: It's better to generalize this section, so I removed host-specific references.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:176]: Fixed.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:209]: I added SGX to the description:
"Thanks for sticking with us. Here is our corrected ecall function that marshals
the parameter correctly for SGX-based enclaves:"

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]: Fixed.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:248]: Added "generally".

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:271]: I took your edits, verbatim.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:273]: I took your edits, verbatim.

**Comment by douglasmaciver** [comment]
@dthaler Thank you for your careful review. 

(Apologies for Markdown text that exceeded the 80 column margin -- I should have caught that.  The downside of that mistake is that correcting it caused a lot of churn, clouding the other small edits that I made to address your minor comments. I just amended the commit instead of adding a new one, given that the delta was so great.)

I would like to continue discussing the two remaining unresolved topics.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I'm beginning to waffle on this. Perhaps we should make the entire Custom Marshalling section explicitly SGX focused?

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: The high-level intent of this topic was to help Enclave developers envision a secure way to handle secrets and Secure Key Release was a prime model. We had not planned on covering the topic in detail here -- we thought we would simply reference existing Open Enclave SDK Secure Key Release documentation, but we could not find any. What we did find was SGX-focused documentation under the Azure Attestation docs: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/security.attestation-readme#attest-sgx-enclave

The text in section 2.2 of our doc is almost verbatim from the above source, so corrections will have to be applied there also.  Would you help me rewrite this short introduction here? I'll submit a doc request for SKR against OE, and submit bugs against the multiple Azure Attestation pages that have this incorrect info.

Comment (dthaler) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I'd prefer to keep the best practices doc generic (not specific to Azure) and cover "best" practices.  The azure doc you link just states it's current practice, it does not assert it is "best" practice as a model to follow, just a model.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:40]:
```suggestion
oeedger8r tool generates. Although this is not generally recommended, it is supported.
To specify this, use `user_check` for the parameter
```

Suggested text is my proposal to make it clear that doing so is not being stated to be a "best practice", but if you DO do it, then this doc covers the best practice on HOW.  Line 61-62 already does so, but this suggestion makes it clear up front for consistency in message.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:146]:
```suggestion
To understand why these functions are _not_ boolean opposites, it helps to
```
(fix grammar mismatch in "this ... are")

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:166]:
```suggestion
functions are not boolean opposites.
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:298]:
```suggestion
Thanks for sticking with us. Here is our updated ecall function that marshals
the parameter correctly for SGX-based enclaves (but is still incorrect for other TEEs
that do not allow enclave code to directly read host memory):
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:295]:
```suggestion
## 1.5 Updated enclave ecall with custom marshaling
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:314]:
```suggestion
    // This line assumes enclave code can directly read host memory.
    captured_data = blob->data;
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:403]: Besides the other problems with this text that I raised, another problem is that the description doesn't match the name.
The name would imply that that a key is only released to an enclave after attestation completes, so it's known to be handed to a trusted client.   In contrast, lines 386-387 instead have the (not yet trusted) enclave _create_ the key and then pass it to the relying party which only _uses_ it if attestation passes.  Thus, it's more like a secure key use than a secure key release.   The only way secure key release would be an accurate name is if you're talking about releasing it to the relying party, but the relying party need not be running in an enclave, and so "secure" doesn't mean the same as what it means in the rest of this article.

**Comment by rs--** [comment]
bors try

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: After some mulling, the proper way to take this guidance is to make it general and hw-agnostic, following the spirit of Open Enclave.  The challenge though, is the potential mixed semantics of the memory models: This is evident in the different memory models of SGX and OP-TEE. When both are considered, it's difficult to choose the proper abstraction: SGX is satisfied by "outside" or "within" the enclave, but with OP-TEE it's "secure" or "nonsecure" -- the memory buffer could be shared, and thus neither necessarily "outside" nor "within" the enclave. 

BTW, I've always been bothered by the OE OP-TEE implementation of `oe_is_outside_enclave()`, and now I'm convinced it's incorrect in ways similar to why the SGX implementations are not boolean opposites. @dthaler, what do you think?
https://github.com/openenclave/openenclave/blob/cd72fd7069488ba6f453c8f5f47bd9fd9a6e6c0d/enclave/core/optee/bounds.c#L33-L36

Let me attempt to generalize the text in light of the comments above.

Comment (dthaler) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Most of the points in the doc apply equally to SGX, OP-TEE, and any other TEEs, so it's good to keep the doc general where possible and just use TEE-specific examples.  For OP_TEE in particular, the discussion in #2139 is relevant to this doc... is_within_enclave() still makes sense as being secure, non-shared.  is_outside_enclave() is more nuanced as discussed in that issue.

Comment (dthaler) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: My personal view is that `oe_is_outside_enclave()` is a confusingly named API to begin with (for both SGX and OP-TEE) for the reasons articulated in the PR and in #2139.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: For the scope of this best practice guidance, I will introduce the Secure Key Release design pattern and exclude implementation details. When SKR for Open Enclave is documented, a link can be included here.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:299]: When I generalize the memory bounds discussion, I'm not going mention "host", which, as you have pointed out, is SGX-centric. Instead, memory buffers will be referred generically as either "trusted" or "untrusted" from the point of view of the enclave. Whether memory is strictly "within" or "outside" the enclave doesn't apply to all hardware implementations. (It is unfortunate that these function names don't map well to this generalized Open Enclave memory model.) Data in buffers that are "untrusted" -- or "outside" -- the enclave must be handled with care. The untrusted buffer contents may be maliciously constructed and may change from read to read.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:296]: I will adopt "updated", but I have to change the table of contents too, so I'll handle it my broader update.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:315]: I'm going to drop the "host" concept. See my response : https://github.com/openenclave/openenclave/pull/4178#discussion_r675225002

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:404]: I agree. Trying to use existing "Secure Key Release" material did not work for this context. I removed the offending material and briefly sketched how attestation could be used with secure data services to manage secrets. It's difficult to go into detail and remain cloud service provider agnostic.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Please see the related conversation here: https://github.com/openenclave/openenclave/pull/4178#discussion_r675261490

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:404]: Lines 387-404 are "the offending material".  Are you removing those lines in another commit?

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: 71be4bcb generalizes the text covering memory models.  I'd appreciate your continued review @dthaler.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: fixed in  71be4bcb,

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:317]: Addressed in  71be4bcb.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Addressed in 71be4bcb. I removed design details of the application secrets section. I can see the benefit of a detailed description of how to perform Secure Key Release (for example) in a cloud-agnostic way, but it's beyond the scope of this doc. I would appreciate your review @dthaler.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:161]:
```suggestion
return `false`, which should _not_ be interpreted as "the buffer is strictly outside the
```
for consistency with line 114.  Or use "entirely" for consistency with line 229.  For bonus points, make 114 and 229 use the same word.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:68]:
```suggestion
correcting this code as we move through the material. The code also assumes that the enclave
can directly read the blob memory, which is true for SGX but may not be true for other TEEs.  The general impetus for
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (dthaler) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:317]: The word was dropped but not the concept.  See suggested fix in line 68

Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:317]: The concept shifted from "host" memory to "shared" memory. OP-TEE uses "shared" memory also, it just happens to be mapped into its address space and marked as TEE_MEMORY_ACCESS_NONSECURE.  (As I understand it -- I'm not an OP-TEE authority.) For OP-TEE, the fact that it is "within" the enclave is what makes generically discussing the OE memory model difficult, especially with functions that focus on "within" and "outside" by name.

Comment (dthaler) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:317]: Some TEEs do not support shared memory at all.   OP-TEE requires the host side to opt-in to sharing the memory with the enclave.   I believe for user-check this is not done for OP-TEE, so anything marked user-check cannot be dereferenced unless is_inside_enclave is true.

Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:317]: Okay, I will take the suggested edits. Thanks for your help Dave.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:162]: Again, good catch.  I adopted "strictly" throughout the doc to align with the Open Enclave API documentation.

Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I also reduced "memory region" to "memory".

**Review by dthaler — APPROVED** [review]

**Comment by douglasmaciver** [comment]
bors try

**Comment by rs--** [comment]
@radhikaj can we get someone to have a look at this PR for the second review? Thanks

**Comment by douglasmaciver** [comment]
@radhikaj I nominate @mingweishih as the second reviewer. He contributed to an early version of the doc.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (mingweishih) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:5]:
```suggestion
1. [Best practices for interface custom marshaling](#1-best-practices-for-interface-custom-marshaling)
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (mingweishih) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:45]:
```suggestion
enclave {
```

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (mingweishih) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:151]:
```suggestion
[memory.c](/enclave/core/sgx/memory.c).
```

**Review by mingweishih — APPROVED** [review]
LGTM. Thanks for putting the information together. Approve with little suggestions on the format.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Fixed in latest commit.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Fixed in latest commit.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Fixed in latest commit.

**Pull request merged**
